src/Estimators/src/InvariantEKFBaseEstimator.cpp

```diff
@@ -0,0 +1,918 @@
+        << std::endl;
+        return false;
+    }
+
+    // setup initial state standard deviations
+    auto priorDevHandle = handle->getGroup("PriorsStdDev");
+    if (!setupPriorDevs(priorDevHandle))
+    {
+        std::cerr << "[InvariantEKFBaseEstimator::customInitialization] Could not load prior stddev related parameters."
+        << std::endl;
+        return false;
+    }
+
+    m_pimpl->m_skew.gCross = iDynTree::skew(m_options.accelerationDueToGravity);
+
+    m_pimpl->constuctStateVar(m_priors, m_options.imuBiasEstimationEnabled, m_pimpl->m_P); // construct priors
+    return true;
+}
+
+bool InvariantEKFBaseEstimator::resetEstimator(const FloatingBaseEstimators::InternalState& newState,
+                                               const FloatingBaseEstimators::StateStdDev& newPriorDev)
+{
+    m_state = newState;
+    m_stateStdDev = newPriorDev;
+    m_priors = newPriorDev;
+
+    m_pimpl->constuctStateVar(m_priors, m_options.imuBiasEstimationEnabled, m_pimpl->m_P); // construct priors
+    return true;
+}
+
+bool InvariantEKFBaseEstimator::resetEstimator(const FloatingBaseEstimators::InternalState& newState,
+                                               const FloatingBaseEstimators::StateStdDev& newPriorDev,
+                                               const FloatingBaseEstimators::SensorsStdDev& newSensorsDev)
+{
+    m_sensorsDev = newSensorsDev;
+    resetEstimator(newState, newPriorDev);
+    return true;
+}
+
+bool InvariantEKFBaseEstimator::predictState(const FloatingBaseEstimators::Measurements& meas,
+                                             const double& dt)
+{
+    m_pimpl->calcSkewSymAtCurrenState(m_statePrev, m_pimpl->m_skew); // compute skews at priori state
+
+    // update foot position predictions with previous measures
+    iDynTree::JointPosDoubleArray jointPos(meas.encoders.size());
+    iDynTree::toEigen(jointPos) = meas.encoders;
+    iDynTree::Transform IMU_H_LF, IMU_H_RF;
+    if (!m_modelComp.getIMU_H_feet(jointPos, IMU_H_LF, IMU_H_RF))
+    {
+        return false;
+    }
+
+    // m_state is now predicted state after this function call
+    if (!m_pimpl->propagateStates(meas, dt, m_options.accelerationDueToGravity, IMU_H_RF, IMU_H_LF, m_state))
+    {
+        return false;
+    }
+
+    Eigen::MatrixXd Fc, Qc, Lc;
+    m_pimpl->calcFc(m_statePrev, m_pimpl->m_skew, m_options.imuBiasEstimationEnabled, Fc); // compute Fc at priori state
+    m_pimpl->calcQc(m_statePrev, m_sensorsDev, meas,  m_options.imuBiasEstimationEnabled, Qc); // compute Qc at priori state and previous measure
+    m_pimpl->calcLc(m_statePrev, m_pimpl->m_skew, m_options.imuBiasEstimationEnabled, Lc); // compute Lc at priori state
+
+    // discretize linearized dynamics and propagate covariance
+    size_t dims = Fc.rows();
+    Eigen::MatrixXd I = Eigen::MatrixXd::Identity(dims, dims);
+
+    Eigen::MatrixXd Fk = I + (Fc*dt);
+    Eigen::MatrixXd Qk = (Fk*Lc*Qc*(Lc.transpose())*(Fk.transpose()))*dt;
```

Comment:
Do you think we may avoid continuous memory allocation if `I` ,`Fk` and `Qk` are class members?

Reply:
Actually, here the catch is that these matrices do not have a fixed dimension throughout the run time. The dimensions vary if bias estimation is toggled during run time. 

In this case, storing the matrices as class members means that we need to modify the mechanics of current implementation in a way that these matrix dimensions are updated at every step. It is not difficult to implement, however, in any case, we need to be reallocating the memory during the resize. That's why I chose this way instead.

Correct me if my rationale is not good enough.

Reply:
I think that some resizing from time to time are better than allocating memory all the time. Can you provide some more insights on why a continuous resize may be necessary?

Reply:
Continuous resize won't be necessary. Just at events of bias estimation toggle. I can make the relevant changes.

Here, we do not even consider arbitrary number of contacts (so we don't need to continuously resize the matrices.)

Reply:
Done in https://github.com/dic-iit/bipedal-locomotion-framework/pull/85/commits/6abd67857f238dc60818b25e497be228b62ab230